### ISSUE TITLE
Trakt not marking episodes as watched

### ISSUE THREAD

**Issue by hluaces** [123Venom/plugin.video.venom#96]
Hello there.

Let me start by saying that I'm a big fan of your work. If there's anyway to donate to your efforts, please let me know because I'd like to.

With that being said, I'm experiencing some issues when finishing shows with Venom. After they are done they are not marked as "watched" on Trakt.tv, which means that anytime that I use the "_Progress_" menu to resume watching a tv show the list shows already seen episodes.

I'm aware that I'm not giving away too much information, but that's simply because I don't know what information might be needed to help.

Could you please let me know what kind of information, logs or screenshots would be needed to try and investigate this?

I'm using Kodi 19 and the latest version of Venom.

**Comment by 123Venom** [comment]
There were a few issues that was in the initial launch of the window xml version of Source Select due to ID's missing during transfer to the player for marking watched while watching.  This was fixed in v5.1.9 and I've yet to hear another issue they still are not working.  I have 3 devices here from win10 to Android/Shield and none have an issue now.

What device?
watching from inside Venom or widget?
Can you manually mark the item watched? check both inside Venom and widget

**Comment by GodsLikelyMe** [comment]
> There were a few issues that were in the initial launch of the window xml version of Source Select due to ID's missing during transfer to the player for marking watched while watching. This was fixed in v5.1.9 and I've yet to hear another issue they still are not working. I have 3 devices here from win10 to Android/Shield and none have an issue now.
> 
> What device?
> watching from inside Venom or widget?
> Can you manually mark the item watched? check both inside Venom and widget

I'm also having this issue with 5.1.9.  Only way to mark them watched is by the context menu. I have watched episodes from within Venom itself and from TMDB helper widgets and gotten the same result. I'm on a FireTV. I've gone thru an entire season and have had none of the episodes automatically marked as watched.

**Comment by hluaces** [comment]
In my case:

* Kodi 19.1
* Venom 5.1.9
* The device is a Chromecast (the model with Android TV)
* I can mark them as watched from the episode list and it works as expected.
* I have only watched episodes from within Venom, couldn't try the widget.

Is there any log/debug information that can help to find this problem?

**Comment by 123Venom** [comment]
@hluaces you can turn on the Venom logging under the **Debugging** category.  Set log level to **Debug**  as this will log extra info that might help me track it.  Also set **Logging Location** to **venom.log**.  That writes the data to a separate log so I'm not hunting through all of kodi's log data.  Toss up a link to the log and I'll see if there's anything vital in there to help me pin point this.  I've not been able to duplicate it here on 3 devices...but..I'll say that last night I spent a few hours with a user trying to solve his same problem but he's using library integration and when he finished watched an item from library Venom did not mark it also inside the addon, and it should.  That I was able to track down to an id that went missing due to pulling Kodi library database info and merging the meta.  You're not using library integration by chance?  The whole meta process changed significantly for the new window xml source select so that required info was provided and not seeing blank sections in the skin even if it was a library scrape.  I'm pretty confident the issue for you is in there, could be pure luck that what I fixed for him last night might very well be related.  I may toss a zip up in the dev branch of the fix for you to test and see if that's also your issue.

**Comment by 123Venom** [comment]
[https://github.com/123Venom/zips/tree/dev](https://github.com/123Venom/zips/tree/dev)

v5.1.9.6 test version available.  If you could test and see if that happens to resolve your issue.  If not I'll need the log to start chasing it.

**Comment by GodsLikelyMe** [comment]
> https://github.com/123Venom/zips/tree/dev
> 
> v5.1.9.6 test version available. If you could test and see if that happens to resolve your issue. If not I'll need the log to start chasing it.

I don't use library integration, but I went ahead and tried the zip you posted. Unfortunately it didn't change anything. Episodes are still not being marked as watched.  I grabbed a log with the instructions you provide, but I don't see anything trakt related in it. 

Log
https://paste.ubuntu.com/p/6YM55RYd9w/

**Comment by 123Venom** [comment]
There's nothing indicative of the issue in the log unfortunately.  Do you use AutoPlay or Source Select?  At this point about the only way I'll find what's happening on your end is to make a custom version that logs the ID's and the Trakt post to mark and see what's happening.  I'll see about getting that done and uplaoded if your game to run tests.  You are using the setting for Indicators set to **Trakt** and not **Local** Indicators correct?

**Comment by GodsLikelyMe** [comment]
> There's nothing indicative of the issue in the log unfortunately. Do you use AutoPlay or Source Select? At this point about the only way I'll find what's happening on your end is to make a custom version that logs the ID's and the Trakt post to mark and see what's happening. I'll see about getting that done and uplaoded if your game to run tests. You are using the setting for Indicators set to **Trakt** and not **Local** Indicators correct?

I'm up for it. I use both auto for episodes and source select for movies. Neither one marks them. And yes I use trakt indicators

**Comment by 123Venom** [comment]
In order for continual playback of episodes you must have the UpNext integration setting on then?  That setting forces AutoPlay for episodes so `Play from Here` works, which is a Kodi controlled playlist launcher.  That's honestly how the 4 devices in this household are setup.  I'll try to get you a modded version with a bunch of logging so we can track it..stinks I can't replicate it because that makes tracking it more of a pain.  Also the next release I'm removing the UpNext addon.  I've written my own for that so the integration is more consistent.  The UpNext addon goes from working to not working for 3rd party addons to often for my taste so I'm replacing it.

**Comment by GodsLikelyMe** [comment]
> In order for continual playback of episodes you must have the UpNext integration setting on then? That setting forces AutoPlay for episodes so `Play from Here` works, which is a Kodi controlled playlist launcher. That's honestly how the 4 devices in this household are setup. I'll try to get you a modded version with a bunch of logging so we can track it..stinks I can't replicate it because that makes tracking it more of a pain. Also the next release I'm removing the UpNext addon. I've written my own for that so the integration is more consistent. The UpNext addon goes from working to not working for 3rd party addons to often for my taste so I'm replacing it.

Nice and thanks for helping out. It really is strange that you don't have that issue on any device. I even tried uninstalling everything that came with venom and deleting their data and folders to try a fresh install, but it's still the same.

**Comment by 123Venom** [comment]
@GodsLikelyMe I've uploaded another test version with quite a bit of logging so I can track the data flow.  Leave the Venom logging set to **Normal**.  Upload both venom and kodi logs.  Also that version is very preliminary for the UpNext change out so feel free to play with that if you like just keep in mind it's got a ways to go before completion.

**Comment by GodsLikelyMe** [comment]
> @GodsLikelyMe I've uploaded another test version with quite a bit of logging so I can track the data flow. Leave the Venom logging set to **Normal**. Upload both venom and kodi logs. Also that version is very preliminary for the UpNext change out so feel free to play with that if you like just keep in mind it's got a ways to go before completion.

Ok, so not sure what else you did, but the problem is fixed! I tried a few different ways which I'll list.

First I played an episode from inside Venom by just hitting play (not play from here) and it marked on its own when it finished. Then I played 2 episodes using "play from here" and they also both marked as watched. The new UpNext feature worked fine too. The preview looks very nice btw. Next I played the newest Bad Batch episode from my tmdb helper widget using auto play choice in the player. It also marked as watched when it finished. I also wanted to check with movies, so I fast forwarded a movie from within venom and 1 from tmdb helper widget and they also both marked watched. 

1 thing I did notice with the new UpNext is that when I finished watching the newest bad Batch, the preview popped up and wanted to start playing the next episode from the last series I had been watching since there were no more Bad Batch episodes. All in all everything is working great! I'm still adding both logs as you requested. Thank you very much for the help. 

Venom Log
https://paste.ubuntu.com/p/2JsCsBGB7J/

Kodi Log
https://paste.ubuntu.com/p/Yg2pgzDbgX/

**Comment by 123Venom** [comment]
Bout the only thing that comes to mind what I change from that other version is in the **keepAlive()** for the player.  I was having some random issues with the PlayNext new pop up last night and spotted a condition where Kodi callbacks were not occurring.  It might have just bitten you first and we just got lucky that's it..lol.  Your venom log looked good.  The Trakt response data indicated they were marked as watched.  All the meta I reviewed indicated nothing was missing, which was my big concern.  I did spot some errors in you kodi.log about subtitle errors.  Might be something for me to investigate at some point.  I tell people to use **ak4subtitles**.

As far as the PlayNext...ya I told you that's got a WAYS to go yet.  First order was getting the pop up successful.  Today I worked on the window xml format because my desire is to make that as small as possible and be non intrusive to the playing video.  I need to code a playlist **getPosition()** check when it reaches the end of the playlist, which is what you witnessed...just not had the time yet and was not the primary concern yet.  I also have a complete window xml to do for a `Still Watching` so after say 3 episodes a new pop up style will show case you fell asleep...lol.  Work in Progress
Keep an eyes on it and let me know if you spot this again.

**Comment by hluaces** [comment]
Just for clarification: I assume that whatever was causing the problem was fixed in one of the test versions, so I assume those changes are gonna get pushed at some point in the future. Did I get it right?

**Comment by GodsLikelyMe** [comment]
@123Venom Will do. If anything wonky happens, I'll make sure to post some logs. Thanks again for all the work you do. Looking forward to the next update. 

@hluaces Yes. The test version 5.1.9.8 fixed it for me, but as it's a test version, there are still some WIP features like the dev mentioned.

**Comment by 123Venom** [comment]
@GodsLikelyMe I've updated a new test version to dev branch.  I believe I've now completed the PlayNext addition.  There's new settings and a second notification called **Still Watching?**  I added the needed code so when you reach the end of the playlist that final item will not display a pop up when playback ends.

Hopefully after a few days testing if all looks good I'll push official this weekend so the Trakt mark watched issue will also be included.
